src/dune_engine/dep.ml

```diff
@@ -125,6 +127,40 @@ module Fact = struct
           Digest.generic
             (Path.Map.to_list_map files ~f:(fun p d -> (Path.to_string p, d)))
       }
+
+    let empty = lazy (make Path.Map.empty)
+
+    let group ts files =
+      let ts =
+        if Path.Map.is_empty files then
+          ts
+        else
+          make files :: ts
+      in
+      (* Sort and de-dup so that the result is resilient to code changes *)
+      let ts =
+        List.filter_map ts ~f:(fun t ->
+            if Path.Map.is_empty t.files then
+              None
+            else
+              Some (t.digest, t))
+        |> Digest.Map.of_list_reduce ~f:(fun t _ -> t)
+        |> Digest.Map.values
+      in
+      match ts with
+      | [] -> Lazy.force empty
+      | [ t ] -> t
+      | t :: l ->
+        { files =
+            List.fold_left l ~init:t.files ~f:(fun acc t ->
```

Comment:
It seems that this `fold_left` is doing a similar amount of work as we were doing before, or am I missing something?
I thought that to benefit from this grouping we would need to delay the evaluation of `files` until we actually need them, e.g. to construct a sandbox.

Reply:
We still win on the digests, but yeah that's true we could do better. Now that sync memo is gone, I suppose we could safely put that under a lazy.

Reply:
Fair, that sounds useful enough.